### ISSUE TITLE
uv_async_send() inconsistency

### ISSUE THREAD

**Issue by santigimeno** [libuv/libuv#3622]
Not sure if this has been brought up before but calling `uv_async_send()` after having closed the `uv_async_t` handle:
- Succeeds on unices.
- Asserts on Windows: https://github.com/libuv/libuv/blob/v1.x/src/win/async.c#L76.

I guess this inconsistency should be fixed but, does checking `handle->flags` in `uv_async_send()` break the thread-safety of the method?

**Comment by vtjnash** [comment]
Yes, that should require C11 atomics to be a valid assert

**Comment by santigimeno** [comment]
But using atomics for that would require to also use atomics for every operation that modifies `handle->flags`, right?

**Comment by vtjnash** [comment]
Yes. The executed code should be the same, but the compiler annotations could be very annoying to add. It is perhaps better to move this assert to the handlers for `close_cb` and `async_cb`, so that it is checked synchronously instead.

**Comment by vtjnash** [comment]
I wrote a clang pass that can print out the location of every place that would need to be thus annotated (https://github.com/JuliaLang/julia/blob/master/src/clangsa/ImplicitAtomics.cpp), and otherwise fail the build, if we do eventually decide we want to go that path.

**Comment by trevnorris** [comment]
@vtjnash Now that #3683 has landed, can this issue be closed?

**Comment by vtjnash** [comment]
It is not related to that issue. It was mostly fixed in v2 many years ago, so I haven't cared about it much. But it may need some bits of https://github.com/libuv/libuv/pull/3879 ported from unix to v2 windows to complete the fix.

**Comment by hhugo** [comment]
I think I'm hitting this on windows https://github.com/ocsigen/js_of_ocaml/actions/runs/12812866907/job/35734042391?pr=1799

**Comment by hhugo** [comment]
Are there any work around?

**Comment by hhugo** [comment]
Could be related to https://github.com/nodejs/node/pull/57910 maybe
